pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java

```diff
@@ -158,6 +158,49 @@ protected void addOrReplaceSegment(ImmutableSegmentImpl segment, ThreadSafeMutab
     }
   }
 
+  /**
+   * <li> When the replacing segment and current segment are of {@link LLCSegmentName} then the PK should resolve to
+   * row in segment with higher sequence id.
+   * <li> When the replacing segment and current segment are of {@link UploadedRealtimeSegmentName} then the PK
+   * should resolve to row in segment with higher creation time followed by sequence id.
+   * <li> For other cases resolve based on creation time of segment. In case the creation time is same, give
+   * preference to an uplaoded segment. A segment which is not LLCSegment can be assumed to be uploaded segment and
+   * is given preference.
+   *
+   * @param segmentName replacing segment name
+   * @param currentSegmentName current segment name having the record for the given primary key
+   * @param segmentCreationTimeMs replacing segment creation time
+   * @param currentSegmentCreationTimeMs current segment creation time
+   * @return true if the record in replacing segment should replace the record in current segment
+   */
+  private boolean shouldReplaceOnComparisonTie(String segmentName, String currentSegmentName,
+      long segmentCreationTimeMs, long currentSegmentCreationTimeMs) {
+
+    LLCSegmentName llcSegmentName = LLCSegmentName.of(segmentName);
+    LLCSegmentName currentLLCSegmentName = LLCSegmentName.of(currentSegmentName);
+    if (llcSegmentName != null && currentLLCSegmentName != null) {
+      return llcSegmentName.getSequenceNumber() > currentLLCSegmentName.getSequenceNumber();
+    }
+
+    int creationTimeComparisonRes = Long.compare(segmentCreationTimeMs, currentSegmentCreationTimeMs);
+
+    UploadedRealtimeSegmentName uploadedSegmentName = UploadedRealtimeSegmentName.of(segmentName);
+    UploadedRealtimeSegmentName currentUploadedSegmentName = UploadedRealtimeSegmentName.of(currentSegmentName);
+    if (uploadedSegmentName != null && currentUploadedSegmentName != null) {
+      if (creationTimeComparisonRes == 0) {
+        return uploadedSegmentName.getSequenceId() > currentUploadedSegmentName.getSequenceId();
+      } else {
+        return creationTimeComparisonRes > 0;
+      }
+    }
+
+    if (creationTimeComparisonRes == 0) {
+      return llcSegmentName == null || uploadedSegmentName != null;
```

Comment:
nit: comment that `uploadedSegmentName != null` is to favor the segment with formatted name